### DISCUSSION TITLE
RequestContext not working with NestJs service wrapper

### DISCUSSION THREAD

**Discussion by xenoterracide** [mikro-orm/mikro-orm#3929] [Q&A]
```ts
@Injectable()
export class StorefrontProductApplicationService
  implements Pick<StorefrontProductRepository, 'findOneOrFail'>
{
  constructor(
    private readonly repo: StorefrontProductRepository,
    protected readonly orm: MikroORM,
  ) {}

  @UseRequestContext()
  findOneOrFail(
    storefrontId: StorefrontId,
    productId: ProductId,
  ): Promise<StorefrontProduct> {
    return this.repo.findOneOrFail(storefrontId, productId);
  }
}
```

repo here is actually another service that wraps around an `EntityRepository`.

I'm not sure why this doesn't work though. I still get

```
 ● subscription for storefront product › storefrontProductUpdated Subscription - Storefront Product Changed Event for a given Storefront

    An error occurred while executing step "an open storefrontProductUpdated Subscription for a storefrontId "8cf9e353-3e65-45db-b73c-fa8493d58a76"": Using global EntityManager instance methods for context specific actions is disallowed. If you need to work with the global instance's identity map, use `allowGlobalContext` configuration option or `fork()` instead.
```

**Comment by B4nan** [comment]
First of all, `@UseRequestContext()` decorator always creates new (fresh) context, so it's not a very good idea to use that on a repository level. You want one context for request/event/whatever handler, this would create a new one for each method call. `flush` on the parent would be a no-op - you wouldn't have any way to flush changes on those entities.

You should be using middlewares with the `RequestContext` helper instead, or at least put the decorator on controller methods/endpoint handlers, but the decorator approach is meant to be used for CRON jobs and stuff like that, you don't need that if you have middlewares available.

**Reply by xenoterracide** [reply]
well, the problem comes from using rxjs, and an event bus, which has no request context either.

```ts
 storefrontProductUpdated(
    @Args('filter')
    { storefrontId }: StorefrontProductChangedFilter,
  ): AsyncIterableIterator<unknown> {
    return observableToAsyncIterable(
      this.storefrontProductChangedEvent$.pipe(
        filter(({ payload }) => payload.storefrontId === storefrontId),
        concatMap(async ({ payload }) => {
          const aggregate = await RequestContext.createAsync(
            this.orm.em,
            async () =>
              await this.storefrontProductRepository.findOneOrFail(
                payload.storefrontId,
                payload.productId,
              ),
          );
          return aggregate;
        }),
        map((entity) => {
          return new StorefrontProductNode(entity);
        }),
      ),
    );
  }
```

I assumed that wrapping `await this.storefrontProductRepository.findOneOrFail(` with a method annotated with `@UseRequestContext()` would work, instead I had to use `RequestContext.createAsync(this.orm.em`
